### PR TITLE
Implement Array#pack('U')

### PR DESCRIPTION
**Pull request by Mogztter**
```
2.4.1 :001 > [49, 50, 51].pack('U*')
 => "123" 
```

```
bundle exec opal -e "require 'corelib/array/pack'; p [49, 50, 51].pack('U*')"  
"123"
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCodePoint

//cc @iliabylich :wink:

### CODE REVIEW

**Comment by iliabylich** [comment]
Could you include a spec for `u` format (`ruby/core/array/pack/u_spec`) in `spec/ruby_specs` file?

**Comment by Mogztter** [comment]
> Could you include a spec for u format (ruby/core/array/pack/u_spec) in spec/ruby_specs file?

Done.

File: spec/filters/bugs/pack_unpack.rb [hunk 1]
```diff
@@ -82,4 +82,29 @@
...
   fails "Array#pack with format 'l' with modifier '>' ignores NULL bytes between directives" # RuntimeError: Unsupported pack directive "l>" (no chunk reader defined)
   fails "Array#pack with format 'l' with modifier '>' ignores spaces between directives" # RuntimeError: Unsupported pack directive "l>" (no chunk reader defined)
+  fails "Array#pack with format 'U' raises a TypeError if #to_int does not return an Integer" # Expected TypeError but no exception was raised ("\u0005" was returned)
+  fails "Array#pack with format 'U' sets the output string to UTF-8 encoding" # Expected #<Encoding:UTF-8> to be computed by "\u0000".encoding (computed #<Encoding:UTF-16LE> instead)
+  fails "Array#pack with format 'U' calls #to_str to coerce the directives string" # RuntimeError: Unsupported pack directive "x" (no chunk reader defined)
+  fails "Array#pack with format 'U' encodes values larger than UTF-8 max codepoints" # Exception: Invalid code point 1114112
+  fails "Array#pack with format 'U' raises a RangeError if passed a negative number" # Exception: Invalid code point -1
```
Comment (Mogztter) [inline spec/filters/bugs/pack_unpack.rb]: From the documentation:

> A RangeError is thrown if an invalid Unicode code point is given (e.g. "RangeError: NaN is not a valid code point").

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/fromCodePoint

```
String.fromCodePoint('_');      // RangeError
String.fromCodePoint(Infinity); // RangeError
String.fromCodePoint(-1);       // RangeError
String.fromCodePoint(3.14);     // RangeError
String.fromCodePoint(3e-2);     // RangeError
String.fromCodePoint(NaN);      // RangeError
```

File: spec/filters/bugs/pack_unpack.rb [hunk 2]
```diff
@@ -82,4 +82,29 @@
...
+  fails "Array#pack with format 'U' raises a RangeError if passed a number larger than an unsigned 32-bit integer" # Exception: Invalid code point 4294967296
```
Comment (Mogztter) [inline spec/filters/bugs/pack_unpack.rb]: Same here

File: spec/filters/bugs/pack_unpack.rb (hunk 1 above)
Comment (iliabylich) [reply spec/filters/bugs/pack_unpack.rb]: Yes, please

File: spec/filters/bugs/pack_unpack.rb [hunk 3]
```diff
@@ -82,4 +82,29 @@
   fails "Array#pack with format 'l' with modifier '>' encodes the number of array elements specified by the count modifier" # RuntimeError: Unsupported pack directive "l>" (no chunk reader defined)
   fails "Array#pack with format 'l' with modifier '>' ignores NULL bytes between directives" # RuntimeError: Unsupported pack directive "l>" (no chunk reader defined)
   fails "Array#pack with format 'l' with modifier '>' ignores spaces between directives" # RuntimeError: Unsupported pack directive "l>" (no chunk reader defined)
+  fails "Array#pack with format 'U' raises a TypeError if #to_int does not return an Integer" # Expected TypeError but no exception was raised ("\u0005" was returned)
```
Comment (iliabylich) [inline spec/filters/bugs/pack_unpack.rb]: This spec says that there must be a `Opal.coerce_to! :to_int` call on each element of the array, could you add it please? Just for reference: there are a lot of such calls in corelib.

File: spec/filters/bugs/pack_unpack.rb [hunk 4]
```diff
@@ -82,4 +82,29 @@
...
+  fails "Array#pack with format 'U' sets the output string to UTF-8 encoding" # Expected #<Encoding:UTF-8> to be computed by "\u0000".encoding (computed #<Encoding:UTF-16LE> instead)
+  fails "Array#pack with format 'U' calls #to_str to coerce the directives string" # RuntimeError: Unsupported pack directive "x" (no chunk reader defined)
+  fails "Array#pack with format 'U' encodes values larger than UTF-8 max codepoints" # Exception: Invalid code point 1114112
+  fails "Array#pack with format 'U' raises a RangeError if passed a negative number" # Exception: Invalid code point -1
+  fails "Array#pack with format 'U' raises a RangeError if passed a number larger than an unsigned 32-bit integer" # Exception: Invalid code point 4294967296
+  fails "Array#pack with format 'U' taints the output string if the format string is tainted" # RuntimeError: Unsupported pack directive "x" (no chunk reader defined)
```
Comment (iliabylich) [inline spec/filters/bugs/pack_unpack.rb]: Could you move this spec to `spec/filters/unsupported/taint.rb` please?

**Comment by Mogztter** [comment]
Thanks for your review @iliabylich 
I will be away for the next few days. I will take care of it next week.

With a little advance I wish you an happy new year :tada: !

File: spec/filters/bugs/pack_unpack.rb (hunk 3 above)
Comment (Mogztter) [reply spec/filters/bugs/pack_unpack.rb]: I think I'm already calling the `ToInt` function on each element.

File: spec/filters/bugs/pack_unpack.rb (hunk 4 above)
Comment (Mogztter) [reply spec/filters/bugs/pack_unpack.rb]: Done

File: spec/filters/bugs/pack_unpack.rb (hunk 1 above)
Comment (Mogztter) [reply spec/filters/bugs/pack_unpack.rb]: Done

**Comment by iliabylich** [comment]
Thanks!

**Pull request merged**